File: backend/app/api/v1/game.py
"""Game API routes - quiz generation, answering, session management."""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.schemas.game import (
    AnswerResult,
    AnswerSubmit,
    GenerateRequest,
    QuestionResponse,
    SessionCreateResponse,
    SessionResponse,
    SessionSummary,
)
from app.schemas.memo import ApiResponse
from app.services import game_service

router = APIRouter(prefix="/game", tags=["game"])


@router.post("/generate", response_model=ApiResponse[SessionCreateResponse])
async def generate_questions(data: GenerateRequest, db: AsyncSession = Depends(get_db)):
    """Generate quiz questions from Wiki content and create a game session."""
    try:
        questions, session = await game_service.generate_questions(
            db, data.wiki_page_id, data.count, data.difficulty, data.model_id
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return ApiResponse(data=SessionCreateResponse(
        session=SessionResponse.model_validate(session),
        questions=[QuestionResponse.model_validate(q) for q in questions],
    ))


@router.post("/answer", response_model=ApiResponse[AnswerResult])
async def submit_answer(
    data: AnswerSubmit,
    session_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
):
    """Submit an answer for a question in a game session."""
    try:
        answer = await game_service.submit_answer(
            db, session_id, data.question_id, data.user_answer
        )
        question = await game_service.get_question(db, data.question_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

    return ApiResponse(data=AnswerResult(
        question_id=answer.question_id,
        user_answer=answer.user_answer,
        is_correct=answer.is_correct,
        correct_answer=question.correct_answer,
        explanation=question.explanation,
    ))


@router.post("/sessions/{session_id}/finish", response_model=ApiResponse[SessionResponse])
async def finish_session(session_id: str, db: AsyncSession = Depends(get_db)):
    """Mark a game session as completed."""
    try:
        session = await game_service.finish_session(db, session_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

    return ApiResponse(data=SessionResponse.model_validate(session))


@router.get("/sessions/{session_id}", response_model=ApiResponse[SessionSummary])
async def get_session_results(session_id: str, db: AsyncSession = Depends(get_db)):
    """Get session results with all answers and explanations."""
    try:
        session, answers = await game_service.get_session_results(db, session_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

    return ApiResponse(data=SessionSummary(
        session=SessionResponse.model_validate(session),
        answers=[AnswerResult(**a) for a in answers],
    ))

File: frontend/src/api/game.ts
import client from './client'

export interface GameQuestion {
  id: string
  wiki_page_id: string | null
  question_type: string
  difficulty: string
  question_text: string
  options: string // JSON string
  source_title: string | null
}

export interface GameSession {
  id: string
  wiki_page_id: string | null
  model_id: string
  status: string
  total_questions: number
  correct_count: number
  created_at: string
  finished_at: string | null
}

export interface SessionCreateData {
  session: GameSession
  questions: GameQuestion[]
}

export interface AnswerResult {
  question_id: string
  user_answer: string
  is_correct: boolean
  correct_answer: string
  explanation: string | null
}

export interface SessionSummary {
  session: GameSession
  answers: AnswerResult[]
}

export async function generateQuestions(params: {
  wiki_page_id?: string
  count?: number
  difficulty?: string
  model_id: string
}): Promise<SessionCreateData> {
  const { data } = await client.post('/game/generate', params)
  return data.data
}

export async function submitAnswer(
  sessionId: string,
  questionId: string,
  userAnswer: string,
): Promise<AnswerResult> {
  const { data } = await client.post(
    `/game/answer?session_id=${sessionId}`,
    { question_id: questionId, user_answer: userAnswer },
  )
  return data.data
}

export async function finishSession(sessionId: string): Promise<GameSession> {
  const { data } = await client.post(`/game/sessions/${sessionId}/finish`)
  return data.data
}

export async function getSessionResults(sessionId: string): Promise<SessionSummary> {
  const { data } = await client.get(`/game/sessions/${sessionId}`)
  return data.data
}
